pulp_2to3_migration/app/plugin/content.py

```diff
@@ -229,13 +231,49 @@ def get_remote_by_importer_id(importer_id):
             if extra_info:
                 future_relations.update(extra_info)
 
+            if is_multi_artifact:
+                base_path = pulp2content.pulp2_storage_path
+                # This is DistributionTree
+                d_artifacts = []
+                for lce in pulp2lazycatalog:
+                    remote = get_remote_by_importer_id(lce.pulp2_importer_id)
+                    # I am not sure if this is the right way to determine the deferred_download
+                    # for the images
+                    deferred_download = not pulp2content.downloaded
+                    if not remote and deferred_download:
+                        _logger.warn(_('On_demand content cannot be migrated without a remote '
+                                       'pulp2 unit_id: {}'.format(pulp2content.pulp2_id)))
+                        continue
+                    relative_path = os.path.relpath(lce.pulp2_storage_path, start=base_path)
+                    da = DeclarativeArtifact(
+                        artifact=artifact,
+                        url=lce.pulp2_url,
+                        relative_path=relative_path,
+                        remote=remote,
+                        deferred_download=deferred_download)
+                    d_artifacts.append(da)
+
+                # We do this last because we need the remote url which is only found in the LCE
+                # of the image files. There is no LCE for the .treeninfo file itself.
+                relative_path = pulp_2to3_detail_content.relative_path_for_content_artifact
+                da = DeclarativeArtifact(
+                    artifact=Artifact(),
+                    url=urljoin(remote.url, relative_path),
+                    relative_path=relative_path,
+                    remote=self.remote,
+                    deferred_download=False,
+                )
+                d_artifacts.append(da)
+                dc.extra_data = future_relations
+                dc = DeclarativeContent(content=pulp3content, d_artifacts=d_artifacts)
+                await self.put(dc)
```

Comment:
I think this all should be done per each importer_id from lazycatalog.
Imagine you sync the same content through different importers in pulp2.
Right now you are creating artifacts for all the importers and then a treeinfo file for one of them.

Maybe stages API will handle the same artifatcs with different Remotes correctly and the same artifacts will be deduplicated on the way. In this case we need treeinfo files with all the remotes created as well.

Reply:
I've added tracking of Remotes.

Reply:
Should you tag this issue then? https://pulp.plan.io/issues/6375

Reply:
@dralley My changes don't address that issue.